src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -436,138 +432,50 @@ public override void HostProjectChanged(HostProject hostProject)
+        public override void ProjectWorkspaceStateChanged(string projectFilePath, ProjectWorkspaceState workspaceProjectState)
         {
-            if (workspaceProject == null)
-            {
-                throw new ArgumentNullException(nameof(workspaceProject));
-            }
-
-            _foregroundDispatcher.AssertForegroundThread();
-
-            if (!IsSupportedWorkspaceProject(workspaceProject))
-            {
-                return;
-            }
-
-            // The WorkspaceProject initialization never triggers a "Project Add" from out point of view, we
-            // only care if the new WorkspaceProject matches an existing HostProject.
-            if (_projects.TryGetValue(workspaceProject.FilePath, out var entry))
+            if (projectFilePath == null)
             {
-                // If this is a multi-targeting project then we are only interested in a single workspace project. If we already
-                // found one in the past just ignore this one.
-                if (entry.State.WorkspaceProject == null)
-                {
-                    var state = entry.State.WithWorkspaceProject(workspaceProject);
-
-                    var oldSnapshot = entry.GetSnapshot();
-                    entry = new Entry(state);
-                    _projects[workspaceProject.FilePath] = entry;
-                    NotifyListeners(new ProjectChangeEventArgs(oldSnapshot, entry.GetSnapshot(), ProjectChangeKind.ProjectChanged));
-                }
+                throw new ArgumentNullException(nameof(projectFilePath));
             }
-        }
 
-        public override void WorkspaceProjectChanged(Project workspaceProject)
-        {
-            if (workspaceProject == null)
+            if (workspaceProjectState == null)
             {
-                throw new ArgumentNullException(nameof(workspaceProject));
+                throw new ArgumentNullException(nameof(workspaceProjectState));
             }
 
             _foregroundDispatcher.AssertForegroundThread();
 
-            if (!IsSupportedWorkspaceProject(workspaceProject))
-            {
-                return;
-            }
-
-            // We also need to check the projectId here. If this is a multi-targeting project then we are only interested
-            // in a single workspace project. Just use the one that showed up first.
-            if (_projects.TryGetValue(workspaceProject.FilePath, out var entry) &&
-                (entry.State.WorkspaceProject == null || entry.State.WorkspaceProject.Id == workspaceProject.Id) &&
-                (entry.State.WorkspaceProject == null || entry.State.WorkspaceProject.Version.GetNewerVersion(workspaceProject.Version) == workspaceProject.Version))
+            if (_projects.TryGetValue(projectFilePath, out var entry))
             {
-                var state = entry.State.WithWorkspaceProject(workspaceProject);
+                var state = entry.State.WithChangedTagHelpers(workspaceProjectState.TagHelpers);
```

Comment:
this name isn't right. If we add more stuff to `WorkspaceProjectState` then what do we do here?

Reply:
Was gonna have the project state represent each piece of the workspace project state as its own method. That way if we needed to differentiate different PSM change events in the future we could. IE:
```
var state = entry.State.WithChangedTagHelpers(workspaceProjectState.TagHelpers)

if (!object.ReferenceEquals(state, entry.State))
{
    // TagHelpers changed
}

var state = entry.State.WithChangedSomethingNew(workspaceProjectState.SomethingNew)

if (!object.ReferenceEquals(state, entry.State))
{
    // SomethingNew changed
}
```

Buttt then again, I guess we could always just have a generic `ProjectWorkspaceState` changed event. Probably better that way since it ties 1-1 with what's being called into on PSM.